# HG changeset patch
# User Michael Goffioul <michael.goffioul@gmail.com>
# Date 1226493116 0
#      Wed Nov 12 12:31:56 2008 +0000
# Node ID 135c0e7d78020fc0f5eb3344f5e12c4168b2ea83
# Parent  830ad84d26547dded1b027577e08b7326ecdfacb
Extend filesep functionality by allowing to return all valid file separators

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,13 @@
+2008-11-12  Michael Goffioul  <michael.goffioul@gmail.com>
+
+	* dirfcn.cc (Ffilesep): Make it return all file separators when 'all'
+	is given as argument.
+
 2008-11-11  Jaroslav Hajek <highegg@gmail.com>
 
 	* pt-assign.cc: Fix handling of empty cs-lists in assignment LHS.
 
 2008-11-11  John W. Eaton  <jwe@octave.org>
 
 	* lex.l (<MATRIX_START>{S}+): Don't insert separator if next token
 	is a separator.
diff --git a/src/dirfns.cc b/src/dirfns.cc
--- a/src/dirfns.cc
+++ b/src/dirfns.cc
@@ -632,24 +632,44 @@ fnmatch (\"a*b\", @{\"ab\"; \"axyzb\"; \
     print_usage ();
 
   return retval;
 }
 
 DEFUN (filesep, args, ,
   "-*- texinfo -*-\n\
 @deftypefn {Built-in Function} {} filesep ()\n\
+@deftypefnx {Built-in Function} {} filesep ('all')\n\
 Return the system-dependent character used to separate directory names.\n\
+\n\
+If 'all' is given, the function return all valid file separators in\n\
+the form of a string. The list of file separators is system-dependent.\n\
+It is / (forward slash) under UNIX or Mac OS X, / and \\ (forward and\n\
+backward slashes) under Windows.\n\
 @seealso{pathsep, dir, ls}\n\
 @end deftypefn")
 {
   octave_value retval;
 
   if (args.length () == 0)
     retval = file_ops::dir_sep_str ();
+  else if (args.length () == 1)
+    {
+      std::string s = args(0).string_value ();
+
+      if (! error_state)
+	{
+	  if (s == "all")
+	    retval = file_ops::dir_sep_chars ();
+	  else
+	    gripe_wrong_type_arg ("filesep", args(0));
+	}
+      else
+	gripe_wrong_type_arg ("filesep", args(0));
+    }
   else
     print_usage ();
 
   return retval;
 }
 
 DEFUN (pathsep, args, ,
     "-*- texinfo -*-\n\
